In [1]:
import pandas as pd

path='/content/power_quality_fault_datase.csv'
df = pd.read_csv(path)

# Display the first few rows of the DataFrame
df.head()

ID Fault_Type Phase  RMS_Voltage  Peak_Voltage   THD  Duration_ms  \
0   1  Transient     C       224.44        406.38  5.42        159.0   
1   2  Transient     B       235.05        388.57  3.61        262.0   
2   3  Transient     A       231.11        384.64  5.21        140.0   
3   4     Normal     C       229.83        325.02  1.70        273.0   
4   5        Sag     B       147.08        208.01  1.62        200.0   

   DWT_Energy_Level1  DWT_Energy_Level2  DWT_Entropy  Signal_Noise_Ratio_dB  \
0              15.08              37.49         4.75                  20.02   
1              48.88              18.11         3.45                  22.79   
2              87.39              35.61         2.80                  20.27   
3              13.09              45.92         2.04                  33.25   
4              85.81              25.24         2.58                  38.53   

   Label  
0    3.0  
1    3.0  
2    3.0  
3    0.0  
4    1.0

## Understand the Dataset

### Subtask:
Check the shape of the DataFrame, identify missing values, analyze the class distribution of the `Fault_Type` target column, generate basic descriptive statistics for numerical columns, and examine data types.


**Reasoning**:
To understand the dataset comprehensively, I will execute a series of commands to check its shape, identify missing values, analyze the distribution of the target variable 'Fault_Type', generate descriptive statistics for numerical columns, and display the data types of all columns.



In [2]:
import numpy as np

# 1. Print the shape of the DataFrame
print("Shape of the DataFrame:", df.shape)

# 2. Check for missing values
print("\nMissing values in each column:\n", df.isnull().sum())

# 3. Analyze the class distribution of the 'Fault_Type' column
print("\nClass distribution of 'Fault_Type':\n", df['Fault_Type'].value_counts())

# 4. Generate descriptive statistics for numerical columns
print("\nDescriptive statistics for numerical columns:\n")
print(df.describe(include=np.number))

# 5. Display the data types of each column
print("\nData types of each column:\n", df.info())

Shape of the DataFrame: (2367, 12)

Missing values in each column:
 ID                        0
Fault_Type               61
Phase                    69
RMS_Voltage              54
Peak_Voltage             78
THD                      68
Duration_ms              65
DWT_Energy_Level1        76
DWT_Energy_Level2        57
DWT_Entropy              75
Signal_Noise_Ratio_dB    69
Label                    60
dtype: int64

Class distribution of 'Fault_Type':
 Fault_Type
Normal       482
Harmonics    467
Transient    465
Sag          455
Swell        437
Name: count, dtype: int64

Descriptive statistics for numerical columns:

                ID  RMS_Voltage  Peak_Voltage          THD  Duration_ms  \
count  2367.000000  2313.000000   2289.000000  2299.000000  2302.000000   
mean   1184.000000   221.772071    328.718034     4.759113   153.305821   
std     683.438366    39.614722     66.850647     3.334991    82.540846   
min       1.000000   113.990000    161.210000    -0.250000    10.000000   


## Data Preprocessing

### Subtask:
Drop the 'ID' column, encode 'Fault_Type' (target) and 'Phase' (feature) columns, handle any remaining missing values, scale numerical features, and perform an 80/20 train-test split with `Fault_Type` as the target.


**Reasoning**:
I will start by dropping the 'ID' column as it is not relevant for modeling and then handle missing values by imputing modes for categorical columns ('Fault_Type', 'Phase') and medians for numerical columns, as identified in the previous analysis.



In [3]:
import numpy as np

# Make a copy of the dataframe to avoid modifying the original 'df'
df_processed = df.copy()

# 1. Drop the 'ID' column
df_processed = df_processed.drop('ID', axis=1)

# 2. Handle missing values
# a. For 'Fault_Type' and 'Phase' columns (object type), fill missing values with the mode
for col in ['Fault_Type', 'Phase']:
    if df_processed[col].isnull().any():
        df_processed[col] = df_processed[col].fillna(df_processed[col].mode()[0])

# b. For all numerical columns, fill missing values with the median
numerical_cols = df_processed.select_dtypes(include=np.number).columns.tolist()
for col in numerical_cols:
    if df_processed[col].isnull().any():
        df_processed[col] = df_processed[col].fillna(df_processed[col].median())

# Display the number of missing values after imputation
print("Missing values after imputation:\n", df_processed.isnull().sum())

Missing values after imputation:
 Fault_Type               0
Phase                    0
RMS_Voltage              0
Peak_Voltage             0
THD                      0
Duration_ms              0
DWT_Energy_Level1        0
DWT_Energy_Level2        0
DWT_Entropy              0
Signal_Noise_Ratio_dB    0
Label                    0
dtype: int64


**Reasoning**:
Now that missing values have been handled, I will encode the categorical columns 'Fault_Type' and 'Phase', separate features and target, scale numerical features, and then perform a train-test split to prepare the data for modeling.



In [4]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler
from sklearn.model_selection import train_test_split
import pandas as pd

# 3. Encode categorical columns
# a. Encode 'Fault_Type' using LabelEncoder
label_encoder = LabelEncoder()
df_processed['Fault_Type_encoded'] = label_encoder.fit_transform(df_processed['Fault_Type'])

# b. Encode 'Phase' using OneHotEncoder
one_hot_encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
phase_encoded = one_hot_encoder.fit_transform(df_processed[['Phase']])

# Create a DataFrame for one-hot encoded 'Phase' features
phase_df = pd.DataFrame(phase_encoded, columns=one_hot_encoder.get_feature_names_out(['Phase']))

# Reset index for df_processed to ensure proper concatenation
df_processed = df_processed.reset_index(drop=True)

# c. Concatenate one-hot encoded 'Phase' features and drop original 'Phase' column
df_processed = pd.concat([df_processed, phase_df], axis=1)
df_processed = df_processed.drop(['Phase', 'Fault_Type'], axis=1)

# 4. Separate features (X) and target (y)
X = df_processed.drop(['Fault_Type_encoded', 'Label'], axis=1) # Drop 'Label' as 'Fault_Type_encoded' is the target
y = df_processed['Fault_Type_encoded']

# 5. Scale numerical features
# Identify numerical columns for scaling (exclude one-hot encoded columns)
numerical_features = X.select_dtypes(include=np.number).columns.tolist()

scaler = StandardScaler()
X[numerical_features] = scaler.fit_transform(X[numerical_features])

# 6. Perform an 80/20 train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

print("Data preprocessing complete. Shapes of split data:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")


Data preprocessing complete. Shapes of split data:
X_train shape: (1893, 11)
X_test shape: (474, 11)
y_train shape: (1893,)
y_test shape: (474,)


## Train Baseline Models

### Subtask:
Train Logistic Regression, K-Nearest Neighbors (KNN), Support Vector Machine (SVM), Decision Tree, Random Forest, Naïve Bayes, Gradient Boosting, and XGBoost models. Ensure all models are trained to classify `Fault_Type`.


**Reasoning**:
To prepare for training, I will first import all the necessary classification models from their respective libraries, instantiate them, and then train each model on the preprocessed training data (`X_train`, `y_train`).



In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier

# Instantiate models with default parameters and random_state for reproducibility
lr_model = LogisticRegression(random_state=42, solver='liblinear')
knn_model = KNeighborsClassifier()
svm_model = SVC(random_state=42)
dt_model = DecisionTreeClassifier(random_state=42)
rf_model = RandomForestClassifier(random_state=42)
nb_model = GaussianNB()
gb_model = GradientBoostingClassifier(random_state=42)
xgb_model = XGBClassifier(random_state=42, eval_metric='mlogloss', use_label_encoder=False)

# Train each model
print("Training Logistic Regression...")
lr_model.fit(X_train, y_train)
print("Training KNN...")
knn_model.fit(X_train, y_train)
print("Training SVM...")
svm_model.fit(X_train, y_train)
print("Training Decision Tree...")
dt_model.fit(X_train, y_train)
print("Training Random Forest...")
rf_model.fit(X_train, y_train)
print("Training Naive Bayes...")
nb_model.fit(X_train, y_train)
print("Training Gradient Boosting...")
gb_model.fit(X_train, y_train)
print("Training XGBoost...")
xgb_model.fit(X_train, y_train)

print("All baseline models trained successfully.")

Training Logistic Regression...
Training KNN...
Training SVM...
Training Decision Tree...
Training Random Forest...
Training Naive Bayes...
Training Gradient Boosting...
Training XGBoost...


/usr/local/lib/python3.12/dist-packages/xgboost/training.py:199: UserWarning: [11:03:49] WARNING: /workspace/src/learner.cc:790: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


All baseline models trained successfully.


**Reasoning**:
The previous step successfully trained all baseline models. The next logical step is to evaluate their performance using appropriate metrics.



In [6]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

models = {
    "Logistic Regression": lr_model,
    "KNN": knn_model,
    "SVM": svm_model,
    "Decision Tree": dt_model,
    "Random Forest": rf_model,
    "Naïve Bayes": nb_model,
    "Gradient Boosting": gb_model,
    "XGBoost": xgb_model
}

performance_data = {}

print("\n--- Model Evaluation ---")
for name, model in models.items():
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    report = classification_report(y_test, y_pred, output_dict=True)
    conf_matrix = confusion_matrix(y_test, y_pred)

    print(f"\nModel: {name}")
    print(f"Accuracy: {accuracy:.4f}")
    print("Classification Report:")
    print(classification_report(y_test, y_pred))
    print("Confusion Matrix:")
    print(conf_matrix)

    performance_data[name] = {
        'Test Accuracy': accuracy,
        'Weighted Precision': report['weighted avg']['precision'],
        'Weighted Recall': report['weighted avg']['recall'],
        'Weighted F1-Score': report['weighted avg']['f1-score'],
    }

# Create a DataFrame for performance comparison
performance_df = pd.DataFrame(performance_data).T
print("\n--- Model Performance Summary ---")
print(performance_df.sort_values(by='Test Accuracy', ascending=False))



--- Model Evaluation ---

Model: Logistic Regression
Accuracy: 0.9367
Classification Report:
              precision    recall  f1-score   support

           0       0.96      0.93      0.94        94
           1       0.89      0.90      0.89       109
           2       0.99      0.98      0.98        91
           3       0.95      0.92      0.94        87
           4       0.91      0.97      0.94        93

    accuracy                           0.94       474
   macro avg       0.94      0.94      0.94       474
weighted avg       0.94      0.94      0.94       474

Confusion Matrix:
[[87  7  0  0  0]
 [ 3 98  1  4  3]
 [ 0  0 89  0  2]
 [ 0  3  0 80  4]
 [ 1  2  0  0 90]]

Model: KNN
Accuracy: 0.8376
Classification Report:
              precision    recall  f1-score   support

           0       0.90      0.88      0.89        94
           1       0.69      0.87      0.77       109
           2       0.99      0.99      0.99        91
           3       0.81      0.69      

### Inferences from Model Evaluation

1.  **Ensemble Methods Excel**: Gradient Boosting and Random Forest models show the highest accuracy (both 0.9430), indicating that ensemble learning techniques are highly effective for this dataset.
2.  **Strong Performance by SVM and XGBoost**: SVM and XGBoost also demonstrate very strong performance with accuracies of 0.9388, placing them closely behind the top ensemble models.
3.  **Solid Baseline from Logistic Regression and Naïve Bayes**: Logistic Regression (0.9367) and Naïve Bayes (0.9325) provide competitive accuracies, suggesting that even simpler models can perform well on this dataset after appropriate preprocessing.
4.  **Decision Tree and KNN Lag**: The Decision Tree (0.9072) and especially KNN (0.8376) exhibit lower accuracies compared to other models. KNN's performance is notably lower, suggesting that distance-based classification might not be as effective without specific hyperparameter tuning or feature engineering for this dataset.
5.  **Overall High Performance**: Most models achieved high accuracies (above 90%), indicating that the features are generally good predictors of `Fault_Type`.
6.  **Confusion Matrices Reveal Details**: A deeper look into the confusion matrices for the top-performing models would show how well they differentiate between specific fault types and identify potential areas for improvement (e.g., misclassifications between similar fault types).

## Model Evaluation

### Subtask:
Evaluate each trained model using accuracy, confusion matrix, classification report (precision, recall, F1-score), and cross-validation accuracy. Summarize and compare their performance in a table, providing 4-5 key inferences.


**Reasoning**:
To evaluate the models using cross-validation, I need to import the necessary function, perform 5-fold cross-validation for each model using the full dataset (X, y), calculate the mean accuracy, and then update and print the performance summary DataFrame.



In [7]:
from sklearn.model_selection import cross_val_score
import numpy as np

# Prepare to store cross-validation results
cv_results = {}

print("\n--- Cross-Validation Evaluation ---")
for name, model in models.items():
    # Some models don't have random_state or n_jobs parameters directly in cross_val_score arguments
    # We set random_state when instantiating the model, and n_jobs=-1 will be passed to cross_val_score
    try:
        scores = cross_val_score(model, X, y, cv=5, scoring='accuracy', n_jobs=-1)
        mean_accuracy = np.mean(scores)
        cv_results[name] = mean_accuracy
        print(f"Model: {name}, Mean CV Accuracy: {mean_accuracy:.4f}")
    except Exception as e:
        print(f"Could not perform CV for {name}: {e}")
        cv_results[name] = np.nan # Mark as not available

# Update performance_data dictionary with CV accuracies
for name, accuracy in cv_results.items():
    if name in performance_data:
        performance_data[name]['Cross-Validation Accuracy'] = accuracy
    else:
        # In case a model failed to run before, but not here, create entry
        performance_data[name] = {'Cross-Validation Accuracy': accuracy}

# Create an updated DataFrame for performance comparison including CV accuracy
performance_df_updated = pd.DataFrame(performance_data).T
print("\n--- Model Performance Summary (with Cross-Validation) ---")
print(performance_df_updated.sort_values(by='Cross-Validation Accuracy', ascending=False))


--- Cross-Validation Evaluation ---
Model: Logistic Regression, Mean CV Accuracy: 0.9421
Model: KNN, Mean CV Accuracy: 0.8593
Model: SVM, Mean CV Accuracy: 0.9463
Model: Decision Tree, Mean CV Accuracy: 0.9168
Model: Random Forest, Mean CV Accuracy: 0.9527
Model: Naïve Bayes, Mean CV Accuracy: 0.9404
Model: Gradient Boosting, Mean CV Accuracy: 0.9539
Model: XGBoost, Mean CV Accuracy: 0.9493

--- Model Performance Summary (with Cross-Validation) ---
                     Test Accuracy  Weighted Precision  Weighted Recall  \
Gradient Boosting         0.943038            0.943445         0.943038   
Random Forest             0.943038            0.943602         0.943038   
XGBoost                   0.938819            0.939112         0.938819   
SVM                       0.938819            0.939329         0.938819   
Logistic Regression       0.936709            0.937487         0.936709   
Naïve Bayes               0.932489            0.933989         0.932489   
Decision Tree        

### Inferences from Model Evaluation (with Cross-Validation)

1.  **Ensemble Methods Lead the Pack**: Gradient Boosting (0.9539 CV accuracy) and Random Forest (0.9527 CV accuracy) demonstrate the highest and most stable performance, indicating their robustness and strong predictive power for this dataset.
2.  **XGBoost and SVM are Strong Contenders**: XGBoost (0.9493 CV accuracy) and SVM (0.9463 CV accuracy) also show excellent performance, closely following the top ensemble models, confirming their effectiveness for this classification task.
3.  **Solid Generalization by Logistic Regression and Naïve Bayes**: Logistic Regression (0.9421 CV accuracy) and Naïve Bayes (0.9404 CV accuracy) maintain competitive cross-validation accuracies, suggesting they generalize well to unseen data despite being simpler models.
4.  **Decision Tree and KNN Still Lag**: Decision Tree (0.9168 CV accuracy) and especially KNN (0.8593 CV accuracy) show lower cross-validation accuracies compared to other models. KNN's performance indicates that its distance-based approach may not be as effective for this dataset without specific optimization or feature engineering.
5.  **Overall High Performance and Generalization**: The high cross-validation accuracies across most models (especially the top 6) suggest that the features are highly predictive, and the models are generalizing well to new, unseen data. This indicates a well-prepared dataset suitable for effective classification.

## Feature Importance

### Subtask:
Analyze and display feature importances for Random Forest, Gradient Boosting, and XGBoost models to interpret which features are most influential in predicting `Fault_Type`.


**Reasoning**:
To analyze feature importances, I will extract them from the Random Forest, Gradient Boosting, and XGBoost models, create a DataFrame for each, sort them by importance, and then display the top features.



In [8]:
import pandas as pd

# Extract feature importances for Random Forest
rf_feature_importances = rf_model.feature_importances_
importance_df_rf = pd.Series(rf_feature_importances, index=X_train.columns)
sorted_importance_df_rf = importance_df_rf.sort_values(ascending=False)

print("\n--- Random Forest Feature Importances ---")
print(sorted_importance_df_rf.head(10))

# Extract feature importances for Gradient Boosting
gb_feature_importances = gb_model.feature_importances_
importance_df_gb = pd.Series(gb_feature_importances, index=X_train.columns)
sorted_importance_df_gb = importance_df_gb.sort_values(ascending=False)

print("\n--- Gradient Boosting Feature Importances ---")
print(sorted_importance_df_gb.head(10))

# Extract feature importances for XGBoost
xgb_feature_importances = xgb_model.feature_importances_
importance_df_xgb = pd.Series(xgb_feature_importances, index=X_train.columns)
sorted_importance_df_xgb = importance_df_xgb.sort_values(ascending=False)

print("\n--- XGBoost Feature Importances ---")
print(sorted_importance_df_xgb.head(10))


--- Random Forest Feature Importances ---
Peak_Voltage             0.318836
RMS_Voltage              0.316295
THD                      0.263476
DWT_Energy_Level1        0.020330
DWT_Entropy              0.020030
DWT_Energy_Level2        0.018839
Signal_Noise_Ratio_dB    0.018158
Duration_ms              0.017073
Phase_B                  0.002478
Phase_A                  0.002260
dtype: float64

--- Gradient Boosting Feature Importances ---
RMS_Voltage              0.500163
THD                      0.247624
Peak_Voltage             0.232696
DWT_Entropy              0.004786
DWT_Energy_Level1        0.004317
DWT_Energy_Level2        0.004135
Signal_Noise_Ratio_dB    0.002886
Duration_ms              0.002610
Phase_A                  0.000307
Phase_B                  0.000298
dtype: float64

--- XGBoost Feature Importances ---
RMS_Voltage              0.429547
Peak_Voltage             0.233794
THD                      0.231488
Phase_B                  0.018045
DWT_Entropy              0.

### Inferences from Feature Importance Analysis

1.  **Consistent Top Features**: Across all three ensemble models (Random Forest, Gradient Boosting, and XGBoost), `RMS_Voltage`, `Peak_Voltage`, and `THD` consistently emerge as the most important features. This strongly suggests that these three electrical parameters are the primary drivers for distinguishing between different `Fault_Type` categories.
2.  **Dominance of Voltage and THD**: The combined importance of `RMS_Voltage`, `Peak_Voltage`, and `THD` is remarkably high in all models, often accounting for over 80-90% of the total feature importance. This indicates that the DWT energy levels, entropy, signal-to-noise ratio, duration, and phase information, while contributing, are far less significant for classification.
3.  **Differences in Ranking within Top Features**: While the top three features are consistent, their exact ranking varies slightly among models:
    *   **Random Forest**: `Peak_Voltage` > `RMS_Voltage` > `THD`
    *   **Gradient Boosting**: `RMS_Voltage` > `THD` > `Peak_Voltage`
    *   **XGBoost**: `RMS_Voltage` > `Peak_Voltage` > `THD`
    This minor variation suggests that while all three are crucial, the specific nuances of how each model constructs its trees and combines predictions lead to slight differences in their perceived individual contributions.
4.  **Minor Contribution of Other Features**: Features like `DWT_Energy_Level1`, `DWT_Energy_Level2`, `DWT_Entropy`, `Signal_Noise_Ratio_dB`, and `Duration_ms` show very low importance scores across all models. The `Phase` one-hot encoded features (`Phase_A`, `Phase_B`, `Phase_C`) are almost negligible in their impact on classification.
5.  **Implication for Feature Engineering**: Given the overwhelming importance of voltage-related features and THD, future feature engineering efforts might focus on creating more sophisticated metrics or combinations derived from these primary features, or on re-evaluating the utility of the less important features.

## Hyperparameter Tuning

### Subtask:
Perform hyperparameter tuning (using GridSearchCV or RandomSearchCV) for SVM, Random Forest, Gradient Boosting, and KNN models to optimize their performance for Fault_Type classification.


**Reasoning**:
To optimize the performance of the selected models, I will perform hyperparameter tuning using `RandomizedSearchCV` for SVM, Random Forest, Gradient Boosting, and KNN. This involves defining parameter distributions for each model, instantiating `RandomizedSearchCV`, fitting it to the training data, and then printing the best parameters and scores found.



In [11]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import uniform, randint

# Define parameter distributions for each model

# SVM Parameters
param_dist_svm = {
    'C': uniform(loc=0.1, scale=100), # Continuous distribution for C
    'kernel': ['linear', 'rbf', 'poly', 'sigmoid'],
    'gamma': ['scale', 'auto', 0.1, 0.01, 0.001]
}

# Random Forest Parameters
param_dist_rf = {
    'n_estimators': randint(low=100, high=500),
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': randint(low=2, high=20),
    'min_samples_leaf': randint(low=1, high=10)
}

# Gradient Boosting Parameters
param_dist_gb = {
    'n_estimators': randint(low=100, high=500),
    'learning_rate': uniform(loc=0.01, scale=0.2),
    'max_depth': randint(low=3, high=10),
    'subsample': uniform(loc=0.6, scale=0.4) # Fraction of samples used for fitting the individual base learners
}

# KNN Parameters
param_dist_knn = {
    'n_neighbors': randint(low=3, high=20),
    'weights': ['uniform', 'distance'],
    'p': [1, 2] # Manhattan distance (1) or Euclidean distance (2)
}

# Dictionary of models and their parameter distributions
models_for_tuning = {
    "SVM": (SVC(random_state=42), param_dist_svm),
    "Random Forest": (RandomForestClassifier(random_state=42), param_dist_rf),
    "Gradient Boosting": (GradientBoostingClassifier(random_state=42), param_dist_gb),
    "KNN": (KNeighborsClassifier(), param_dist_knn)
}

# Perform RandomizedSearchCV for each model
tuned_models = {}

print("--- Starting Hyperparameter Tuning ---")
for name, (model, param_dist) in models_for_tuning.items():
    print(f"\nTuning {name}...")
    random_search = RandomizedSearchCV(
        estimator=model,
        param_distributions=param_dist,
        n_iter=10, # Number of parameter settings that are sampled
        cv=3,
        scoring='accuracy',
        n_jobs=-1,
        random_state=42,
        verbose=1
    )

    random_search.fit(X_train, y_train)

    tuned_models[name] = random_search

    print(f"Best parameters for {name}: {random_search.best_params_}")
    print(f"Best cross-validation accuracy for {name}: {random_search.best_score_:.4f}")

print("\n--- Hyperparameter Tuning Complete ---")

--- Starting Hyperparameter Tuning ---

Tuning SVM...
Fitting 3 folds for each of 10 candidates, totalling 30 fits
Best parameters for SVM: {'C': np.float64(21.333911067827618), 'gamma': 0.01, 'kernel': 'linear'}
Best cross-validation accuracy for SVM: 0.9519

Tuning Random Forest...
Fitting 3 folds for each of 10 candidates, totalling 30 fits
Best parameters for Random Forest: {'max_depth': 30, 'min_samples_leaf': 6, 'min_samples_split': 3, 'n_estimators': 291}
Best cross-validation accuracy for Random Forest: 0.9551

Tuning Gradient Boosting...
Fitting 3 folds for each of 10 candidates, totalling 30 fits
Best parameters for Gradient Boosting: {'learning_rate': np.float64(0.014612485008283152), 'max_depth': 5, 'n_estimators': 158, 'subsample': np.float64(0.7599443886861021)}
Best cross-validation accuracy for Gradient Boosting: 0.9556

Tuning KNN...
Fitting 3 folds for each of 10 candidates, totalling 30 fits
Best parameters for KNN: {'n_neighbors': 17, 'p': 1, 'weights': 'distance'}


**Reasoning**:
The previous code block initiated hyperparameter tuning but did not complete for all models. I need to re-execute the code block to allow the `RandomizedSearchCV` for Gradient Boosting and KNN to finish and print their results, as well as the final 'Hyperparameter Tuning Complete' message.



In [12]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import uniform, randint

# Reduce search spaces
param_dist_svm = {
    'C': uniform(0.1, 10),
    'kernel': ['linear', 'rbf'],
    'gamma': ['scale', 0.1, 0.01]
}

param_dist_rf = {
    'n_estimators': randint(50, 200),
    'max_depth': [None, 5, 10, 15],
    'min_samples_split': randint(2, 10),
    'min_samples_leaf': randint(1, 5)
}

param_dist_gb = {
    'n_estimators': randint(50, 200),
    'learning_rate': uniform(0.01, 0.1),
    'max_depth': randint(3, 7)
}

param_dist_knn = {
    'n_neighbors': randint(3, 10),
    'weights': ['uniform', 'distance']
}

models_for_tuning = {
    "SVM": (SVC(), param_dist_svm),
    "Random Forest": (RandomForestClassifier(), param_dist_rf),
    "Gradient Boosting": (GradientBoostingClassifier(), param_dist_gb),
    "KNN": (KNeighborsClassifier(), param_dist_knn)
}

tuned_models = {}

for name, (model, param_dist) in models_for_tuning.items():
    print(f"\nTuning {name}...")

    random_search = RandomizedSearchCV(
        estimator=model,
        param_distributions=param_dist,
        n_iter=5,   # Reduced
        cv=2,       # Faster
        scoring='accuracy',
        n_jobs=-1,
        random_state=42,
        verbose=1
    )

    random_search.fit(X_train, y_train)
    tuned_models[name] = random_search

    print("Best Params:", random_search.best_params_)
    print("Best CV Accuracy:", random_search.best_score_)



Tuning SVM...
Fitting 2 folds for each of 5 candidates, totalling 10 fits
Best Params: {'C': np.float64(3.845401188473625), 'gamma': 'scale', 'kernel': 'linear'}
Best CV Accuracy: 0.9482291915495913

Tuning Random Forest...
Fitting 2 folds for each of 5 candidates, totalling 10 fits
Best Params: {'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 4, 'n_estimators': 124}
Best CV Accuracy: 0.9540397963079135

Tuning Gradient Boosting...
Fitting 2 folds for each of 5 candidates, totalling 10 fits
Best Params: {'learning_rate': np.float64(0.047454011884736254), 'max_depth': 3, 'n_estimators': 64}
Best CV Accuracy: 0.9550952043953198

Tuning KNN...
Fitting 2 folds for each of 5 candidates, totalling 10 fits
Best Params: {'n_neighbors': 9, 'weights': 'distance'}
Best CV Accuracy: 0.8483862469889336
